pkg/scheduler/scheduler.go

```diff
@@ -0,0 +1,234 @@
+}
+
+// Register registers and starts new task in the background
+func (s *Scheduler) Register(ctx context.Context, task Task, opts ...Opt) *Definition {
+	id := uuid.New()
+	def := &Definition{
+		scheduler: s,
+		id:        id,
+		group:     DefaultGroup,
+		name:      id.String(),
+		task:      task,
+		interval:  time.Second,
+	}
+	for _, opt := range opts {
+		opt(def)
+	}
+
+	def.logger = newDefinitionLogger(def, s.logger)
+
+	def.startTicker(ctx)
+
+	s.mu.Lock()
+	s.definitions[id] = def
+	s.mu.Unlock()
+
+	return def
+}
+
+// Stop stops all tasks.
+func (s *Scheduler) Stop() {
+	s.StopGroup("")
+}
+
+// StopGroup stops all tasks in the group.
+func (s *Scheduler) StopGroup(group Group) {
+	var selectedDefs []*Definition
+
+	s.mu.RLock()
+
+	// Filter desired definitions
+	for _, def := range s.definitions {
+		// "" is for wildcard i.e. all groups
+		if group == "" || def.group == group {
+			selectedDefs = append(selectedDefs, def)
+		}
+	}
+
+	s.mu.RUnlock()
+
+	if len(selectedDefs) == 0 {
+		return
+	}
+
+	// Stop all selected tasks concurrently
+	var wg sync.WaitGroup
+	wg.Add(len(selectedDefs))
+
+	for _, def := range selectedDefs {
+		go func(def *Definition) {
+			defer wg.Done()
+			def.Stop()
+		}(def)
+	}
+
+	wg.Wait()
+}
+
+// Stop stops the task and offloads it from the scheduler.
+func (d *Definition) Stop() {
+	start := time.Now()
+
+	// delete definition from scheduler
+	defer func() {
+		d.scheduler.mu.Lock()
+		delete(d.scheduler.definitions, d.id)
+		d.scheduler.mu.Unlock()
+
+		timeTakenMS := time.Since(start).Milliseconds()
+		d.logger.Info().Int64("time_taken_ms", timeTakenMS).Msg("Stopped scheduler task")
+	}()
+
+	d.logger.Info().Msg("Stopping scheduler task")
+
+	if d.isIntervalTicker() {
+		d.ticker.StopBlocking()
+		return
+	}
+
+	d.blockChanTicker.Stop()
+}
+
+func (d *Definition) isIntervalTicker() bool {
+	return d.blockChan == nil
+}
+
+func (d *Definition) startTicker(ctx context.Context) {
+	d.logger.Info().Msg("Starting scheduler task")
+
+	if d.isIntervalTicker() {
+		d.ticker = ticker.New(d.interval, d.invokeByInterval, ticker.WithLogger(d.logger, d.name))
+		bg.Work(ctx, d.ticker.Start, bg.WithLogger(d.logger))
+
+		return
+	}
+
+	d.blockChanTicker = newBlockTicker(d.invoke, d.blockChan, d.logger)
+
+	bg.Work(ctx, d.blockChanTicker.Start, bg.WithLogger(d.logger))
+}
+
+// invokeByInterval a ticker.Task wrapper of invoke.
+func (d *Definition) invokeByInterval(ctx context.Context, t *ticker.Ticker) error {
+	if err := d.invoke(ctx); err != nil {
+		d.logger.Error().Err(err).Msg("task failed")
+	}
+
+	if d.intervalUpdater != nil {
+		// noop if interval is not changed
+		t.SetInterval(d.intervalUpdater())
+	}
+
+	return nil
+}
+
+// invoke executes a given Task with logging & telemetry.
+func (d *Definition) invoke(ctx context.Context) error {
+	// skip tick
+	if d.skipper != nil && d.skipper() {
+		return nil
+	}
+
+	d.logger.Debug().Msg("Invoking task")
+
+	err := d.task(ctx)
+
+	// todo metrics (TBD)
```

Comment:
@gartnera would be great to hear your thoughts on this

Reply:
Yeah I think we add one metric then use the task name as a label.

Metrics for:
- count of times run
- error rate
- duration

Reply:
https://github.com/zeta-chain/node/pull/3319/commits/6cc8e435b0b677fe004c50d97f6480d9aaf85326